src/panels/config/integrations/integration-panels/zwave_js/dialog-zwave_js-update-firmware-node.ts

```diff
@@ -0,0 +1,444 @@
+import { HomeAssistant } from "../../../../../types";
+import { ZWaveJSUpdateFirmwareNodeDialogParams } from "./show-dialog-zwave_js-update-firmware-node";
+import {
+  showAlertDialog,
+  showConfirmationDialog,
+} from "../../../../../dialogs/generic/show-dialog-box";
+import { NumberSelector } from "../../../../../data/selector";
+
+@customElement("dialog-zwave_js-update-firmware-node")
+class DialogZWaveJSUpdateFirmwareNode extends LitElement {
+  @property({ attribute: false }) public hass!: HomeAssistant;
+
+  @state() private device?: DeviceRegistryEntry;
+
+  @state() private _uploading = false;
+
+  @state()
+  private _updateFinishedMessage?: ZWaveJSNodeFirmwareUpdateFinishedMessage;
+
+  @state()
+  private _updateProgressMessage?: ZWaveJSNodeFirmwareUpdateProgressMessage;
+
+  @state() private _updateInProgress = false;
+
+  @state() private _firmwareFile?: File;
+
+  @state() private _nodeStatus?: NodeStatus;
+
+  private _subscribedNodeStatus?: Promise<UnsubscribeFunc>;
+
+  private _subscribedNodeFirmwareUpdate?: Promise<UnsubscribeFunc>;
+
+  private _deviceName?: TemplateResult;
+
+  private _target?: number;
+
+  private _firmwareUpdateCapabilities?: ZWaveJSNodeFirmwareUpdateCapabilities;
+
+  public showDialog(params: ZWaveJSUpdateFirmwareNodeDialogParams): void {
+    this._deviceName = html`<strong
+      >${computeDeviceName(params.device, this.hass!)}</strong
+    >`;
+    this.device = params.device;
+    this._fetchData();
+    this._subscribeNodeStatus();
+  }
+
+  public closeDialog(): void {
+    this._unsubscribeNodeFirmwareUpdate();
+    this._unsubscribeNodeStatus();
+    this.device =
+      this._updateProgressMessage =
+      this._updateFinishedMessage =
+      this._firmwareFile =
+      this._nodeStatus =
+      this._target =
+      this._firmwareUpdateCapabilities =
+        undefined;
+    this._uploading = this._updateInProgress = false;
+
+    fireEvent(this, "dialog-closed", { dialog: this.localName });
+  }
+
+  protected render(): TemplateResult {
+    if (
+      !this.device ||
+      !this._nodeStatus ||
+      !this._firmwareUpdateCapabilities ||
+      !this._firmwareUpdateCapabilities.firmware_upgradable ||
+      this._updateInProgress === undefined
+    ) {
+      return html``;
+    }
+
+    const targetSelector: NumberSelector = {
+      number: {
+        step: 1,
+        mode: "box",
+        min: Math.min(...this._firmwareUpdateCapabilities.firmware_targets),
+        max: Math.max(...this._firmwareUpdateCapabilities.firmware_targets),
+      },
+    };
+
+    const beginFirmwareUpdateHTML = html`<ha-file-upload
+        .hass=${this.hass}
+        .uploading=${this._uploading}
+        .icon=${mdiFileUpload}
+        label=${this._firmwareFile?.name ??
+        this.hass.localize(
+          "ui.panel.config.zwave_js.update_firmware.upload_firmware"
+        )}
+        @file-picked=${this._uploadFile}
+      ></ha-file-upload>
+      <ha-selector-number
```

Comment:
Use `ha-form`

Reply:
what's wrong with this schema?
<img width="490" alt="image" src="https://user-images.githubusercontent.com/7243222/174690583-10604101-e3db-44e5-8631-3e5d470fe69c.png">

Reply:
Have you imported ha form? Could possibly be something above this that is making everything red or those vars do not exist

Reply:
you're right I didn't, but the error is still there now:
<img width="693" alt="image" src="https://user-images.githubusercontent.com/7243222/174715191-4744c0f6-7086-4788-bb43-52476a1a2bb3.png">

Reply:
creating a memoized function resolved this issue

Reply:
actually new issue - I wanted this field to be able to accept a blank value to let the driver figure out the right target. But ha-form doesn't appear to let me do that. Any ideas/suggestions?

Reply:
Figured it out!